src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryContainsMethods.cs

```diff
@@ -0,0 +1,223 @@
+                    return;
+
+                //  At this point, we know that the method being invoked is a method called "Contains" that has a boolean return type.
+                //  We also know that the method is being invoked on a property belonging to a type that implements IDictionary`2. We will
+                //  compare the types used to construct IDictionary`2 to the parameter type of the Contains method. 
+
+                ITypeSymbol keyType = constructedDictionaryType.TypeArguments[0];
+                ITypeSymbol valueType = constructedDictionaryType.TypeArguments[1];
+
+                //  We use Parameters.Last() because the first argument could be the key/value collection, depending
+                //  on whether the method is an extension method and whether the language is C# or Visual Basic. 
+
+                ITypeSymbol containsParameterType = containsMethod.Parameters.Last().Type;
+
+                if (property.Name == KeysPropertyName)
+                {
+                    if (!containsParameterType.Equals(keyType, SymbolEqualityComparer.Default))
+                        return;
+
+                    //  Now we search for an accessible ContainsKey method that returns boolean and accepts a single
+                    //  parameter of the type that was substituted for TKey.
+
+                    IMethodSymbol? containsKeyMethod = property.ContainingType.GetMembers(ContainsKeyMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                            x.IsPublic() &&
+                            x.Parameters.Length == 1 &&
+                            x.Parameters[0].Type.Equals(keyType, SymbolEqualityComparer.Default))
+                        .SingleOrDefault();
+
+                    if (containsKeyMethod is null)
+                        return;
+
+                    var diagnostic = Diagnostic.Create(ContainsKeyRule, invocation.Syntax.GetLocation(), property.ContainingType.Name);
+                    context.ReportDiagnostic(diagnostic);
+                }
+                else if (property.Name == ValuesPropertyName)
+                {
+                    if (!containsParameterType.Equals(valueType, SymbolEqualityComparer.Default))
+                        return;
+
+                    //  Now we search for an accessible ContainsValue method that returns boolean and accepts a single
+                    //  parameter of the type that was substituted for TValue.
+
+                    IMethodSymbol? containsValueMethod = property.ContainingType.GetMembers(ContainsValueMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                            x.Parameters.Length == 1 &&
```

Comment:
You have added an extra `IsPublic` check on the key side, is it worth having it there too?

Reply:
Good catch. I'll fix it so it's consistent.